cpp/modmesh/universe/rtree.hpp

```diff
@@ -0,0 +1,722 @@
+#pragma once
+
+/*
+ * Copyright (c) 2025, An-Chi Liu <phy.tiger@gmail.com>
+ *
+ * Redistribution and use in source and binary forms, with or without
+ * modification, are permitted provided that the following conditions are met:
+ *
+ * - Redistributions of source code must retain the above copyright notice,
+ *   this list of conditions and the following disclaimer.
+ * - Redistributions in binary form must reproduce the above copyright notice,
+ *   this list of conditions and the following disclaimer in the documentation
+ *   and/or other materials provided with the distribution.
+ * - Neither the name of the copyright holder nor the names of its contributors
+ *   may be used to endorse or promote products derived from this software
+ *   without specific prior written permission.
+ *
+ * THIS SOFTWARE IS PROVIDED BY THE COPYRIGHT HOLDERS AND CONTRIBUTORS "AS IS"
+ * AND ANY EXPRESS OR IMPLIED WARRANTIES, INCLUDING, BUT NOT LIMITED TO, THE
+ * IMPLIED WARRANTIES OF MERCHANTABILITY AND FITNESS FOR A PARTICULAR PURPOSE
+ * ARE DISCLAIMED. IN NO EVENT SHALL THE COPYRIGHT HOLDER OR CONTRIBUTORS BE
+ * LIABLE FOR ANY DIRECT, INDIRECT, INCIDENTAL, SPECIAL, EXEMPLARY, OR
+ * CONSEQUENTIAL DAMAGES (INCLUDING, BUT NOT LIMITED TO, PROCUREMENT OF
+ * SUBSTITUTE GOODS OR SERVICES; LOSS OF USE, DATA, OR PROFITS; OR BUSINESS
+ * INTERRUPTION) HOWEVER CAUSED AND ON ANY THEORY OF LIABILITY, WHETHER IN
+ * CONTRACT, STRICT LIABILITY, OR TORT (INCLUDING NEGLIGENCE OR OTHERWISE)
+ * ARISING IN ANY WAY OUT OF THE USE OF THIS SOFTWARE, EVEN IF ADVISED OF THE
+ * POSSIBILITY OF SUCH DAMAGE.
+ */
+
+// This library implement R-tree based on "R Trees: A Dynamic Index Structure for Spatial Searchin" by Guttman in 1984
+// https://doi.org/10.1145/971697.602266
+
+#include <modmesh/base.hpp>
+
+#include <algorithm>
+#include <limits>
+#include <memory>
+#include <type_traits>
+#include <vector>
+
+namespace modmesh
+{
+
+/// Bounding box for 2D objects
+/// @tparam T floating-point type
+template <typename T>
+class BoundBox2d : public NumberBase<int32_t, T>
+{
+
+public:
+    using value_type = T;
+
+private:
+    value_type m_min_x;
+    value_type m_min_y;
+    value_type m_max_x;
+    value_type m_max_y;
+
+public:
+
+    /// construct a bounding box with given min and max coordinates
+    BoundBox2d(value_type min_x_in, value_type min_y_in, value_type max_x_in, value_type max_y_in)
+        : m_min_x(min_x_in)
+        , m_min_y(min_y_in)
+        , m_max_x(max_x_in)
+        , m_max_y(max_y_in)
+    {
+    }
+
+    /// construct a bounding box that encloses two bounding boxes
+    BoundBox2d(const BoundBox2d & a, const BoundBox2d & b)
```

Comment:
This is ambiguous, but ok for now.  If we need different behaviors in the future, new arguments can be added.